### PR TITLE
SAI-4620: (FS-specific) pre-filter dynamic docValues fields for field value retrieval

### PR DESCRIPTION
**Pull request by magibney**
This is a bit of a hack, but has a huge impact on latency (reducing for some requests by ~100x). Co-dependent on a corresponding change in UpdateRequestProcessor.

### CODE REVIEW

**Review by justinrsweeney — APPROVED** [review]
LGTM!

File: solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java:621]: is --i this looks good? it will process one less item

Comment (magibney) [reply solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java:621]: yes, because we already processed the first one. It's a bit tricky, true. When you get to this part of the code, `ct` will always be `ct > 1`; so in the lowest possible ct (2), we add the first term before the loop, then i is initialized to 1 (`ct - 1`), then we enter the loop, process the second (final) term; then `--i == 0`, so `--i > 0` evaluates to false and we break out of the loop, which is exactly what we want.

**Review by hiteshk25 — APPROVED** [review]
LGTM

**Comment by magibney** [comment]
will merge this after merging #128.

**Pull request merged**
